### PR TITLE
make find functions sign-sensitive

### LINKED ISSUE #40006: findfirst (findnext) is not sign-sensitive
```
julia> p = ==(Int8(-1))
(::Base.Fix2{typeof(==), Int8}) (generic function with 1 method)

julia> p(0xff)
false

julia> findfirst(p, [0xff])  # ??
1

julia> findfirst(p, b"\xff")  # nothing; this is expected
```

### PR DESCRIPTION
**Pull request by bicycle1885**
Fix #40006

### CODE REVIEW

File: base/strings/search.jl [hunk 1]
```diff
@@ -19,13 +22,13 @@ function findnext(pred::Fix2{<:Union{typeof(isequal),typeof(==)},<:AbstractChar}
...
 
-findnext(pred::Fix2{<:Union{typeof(isequal),typeof(==)},<:Union{Int8,UInt8}}, a::ByteArray, i::Integer) =
+findnext(pred::Fix2{<:Union{typeof(isequal),typeof(==)},T}, a::ByteArray{T}, i::Integer) where T<:Byte =
     nothing_sentinel(_search(a, pred.x, i))
 
-function _search(a::Union{String,ByteArray}, b::Union{Int8,UInt8}, i::Integer = 1)
+function _search(a::Union{String,ByteArray}, b::Byte, i::Integer = 1)
     if i < 1
         throw(BoundsError(a, i))
     end
```
Comment (vtjnash) [inline base/strings/search.jl:34]: Alternatively?
```suggestion
    end
    typemin(eltype(a)) <= b <= typemax(eltype(a)) || return 0
```

Comment (bicycle1885) [reply base/strings/search.jl:34]: Thank you for your suggestion. That looks more elegant. However, I'd like to fix the problem in an upcoming pull request. I'm preparing an overhaul of `findnext` and friends to improve the consistency and performance, which can be found in https://github.com/bicycle1885/StringSearch.jl. This pull request is a stepping stone and the main purpose is to file an issue I found and add some test to make the next one less complicated.

File: base/strings/search.jl [hunk 2]
```diff
@@ -1,5 +1,8 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
+const Byte = Union{Int8,UInt8}
+const ByteArray{T<:Byte} = Vector{T}
```
Comment (JeffBezanson) [inline base/strings/search.jl:4]: This needs to be `Union{Vector{UInt8},Vector{Int8}}`; otherwise it includes `Vector{Union{Int8,UInt8}}`.

Comment (bicycle1885) [reply base/strings/search.jl]: Ah, right. Thank you. Fixed.

**Comment by bicycle1885** [comment]
This pull request is superseded by #40120.

**Pull request closed without merging**